File: src/main/java/com/pavlenko/algorithm/MergeSort.java
package com.pavlenko.algorithm;

import java.util.Arrays;

/**
 * 
 * @author Sergii Pavlenko
 * @since Nov 29, 2015
 *
 */
public class MergeSort {

    public int[] sort(int[] array) {
        if (array == null || array.length <= 1) {
            return array;
        }
        int[] left = Arrays.copyOfRange(array, 0, array.length / 2);
        int[] right = Arrays.copyOfRange(array, array.length / 2, array.length);
        int[] result = merge(sort(left), sort(right));
        return result;
    }

    private int[] merge(int[] left, int[] right) {
        int i = 0, j = 0, k = 0;
        int[] result = new int[left.length + right.length];
        while (k != left.length + right.length) {
            if (i != left.length && (j == right.length || left[i] < right[j])) {
                result[k] = left[i];
                i++;
            } else {
                result[k] = right[j];
                j++;
            }
            k++;
        }

        return result;
    }

}

File: src/main/scala/com/pavlenko/algorithm/MergeSort.scala
package com.pavlenko.algorithm

/**
 * Merge sort algorithm
 *
 * @author sergii.pavlenko
 * @since Jun 27, 2017
 */
object MergeSort {
  def sort(list: List[Long]): List[Long] = {
    val n = list.length / 2
    if (n == 0) list
    else {
      def merge(left: List[Long], right: List[Long]): List[Long] =
        left match {
          case Nil => right
          case x :: left1 =>
            right match {
              case Nil => left
              case y :: right1 =>
                if (x < y) x :: merge(left1, right)
                else y :: merge(left, right1)
            }
        }
      val (left, right) = list splitAt n
      merge(sort(left), sort(right))
    }
  }
}
